### ISSUE TITLE
customfields configuration error/ignored?

### ISSUE THREAD

**Issue by Khinsu** [logfellow/logstash-logback-encoder#168]
Documentation example, produce `com.fasterxml.jackson.core.JsonParseException`: 

`<customFields>{"appname":"myWebservice","roles":["customerorder","auth"],"buildinfo":{"version":"Version 0.1.0-SNAPSHOT","lastcommit":"75473700d5befa953c45f630c6d9105413c16fe1"}}</customFields>`
the following josn will be excepted but will not appear in output:
`"appname":"myWebservice","roles":["customerorder","auth"],"buildinfo":{"version":"Version 0.1.0-SNAPSHOT","lastcommit":"75473700d5befa953c45f630c6d9105413c16fe1"`

setting the customfields programmatically is working fine.

**Comment by philsttr** [comment]
Hi @Khinsu,

This is caused by a bug in logback.... [LOGBACK-1101](http://jira.qos.ch/browse/LOGBACK-1101)

Please vote for that issue on logback's jira site if you'd like to see it fixed.

**Comment by lonerzzz** [comment]
The logback bug seems to be for nested fields in my reading. I am simply trying to do the following without any nested fields:
`<customFields>{"appname":"myWebservice"}</customFields>`

I have added it as a child of the appender and as a child of the encoder with no luck.

Is there a need for a separate bug in logback or is this possibley a bug in the encoder?

**Comment by philsttr** [comment]
Correct.  The logback bug only applies to nested JSON objects.

I just tested the following configuration

```
<appender class="ch.qos.logback.core.ConsoleAppender">
    <encoder class="net.logstash.logback.encoder.LogstashEncoder">
        <customFields>{"appname":"myWebservice"}</customFields>
    </encoder>
</appender>
```

with this Main class...

```
public class Main {
    private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
    
    public static void main(String[] args) throws InterruptedException, IOException {
        LOGGER.info("Some message");
        ((LoggerContext) LoggerFactory.getILoggerFactory()).stop();
    }
}
```

and received this output:

```
{"@timestamp":"2016-07-20T22:18:53.255-07:00","@version":1,"message":"Some message","logger_name":"net.logstash.logback.Main","thread_name":"main","level":"INFO","level_value":20000,"appname":"myWebservice"}
```

So, this appears to be working for me.

Can you provide the exact configuration you are using?

Can you try outputting to the console appender as I have done to examine the encoder output directly?

**Comment by philsttr** [comment]
Note that [LOGBACK-1101](http://jira.qos.ch/browse/LOGBACK-1101) has been marked as fixed in logback 1.3.0-alpha3

**Comment by araj28** [comment]
<appender class="ch.qos.logback.core.ConsoleAppender">
    <encoder class="net.logstash.logback.encoder.LogstashEncoder">
        <customFields>{"appname":"myWebservice"}</customFields>
    </encoder>
</appender>


Is there any way to restrict data type for custom fields to a data type?

**Comment by brenuart** [comment]
@Khinsu Are you still experiencing problems with this?

**Comment by Qi77Qi** [comment]
hi, think I'm running into the same issue

I have the following in logback.xml
```
    <appender name="console" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LogstashEncoder">
            <customFields>
                {"serviceContext":{"service":"foo"}}
            </customFields>
        </encoder>
    </appender>
```

When using `logstash-logback-encoder 7.0.1`, I'm getting the follow error
```
02:17:46,889 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [net.logstash.logback.fieldnames.LogstashFieldNames] for [fieldNames] property
02:17:47,133 |-ERROR in net.logstash.logback.composite.GlobalCustomFieldsJsonProvider@4d5b181d - [customFields] is not a valid JSON object com.fasterxml.jackson.core.io.JsonEOFException: Unexpected end-of-input: expected close marker for Object (start marker at [Source: (String)"{"serviceContext":{"service":"leonardo"}"; line: 1, column: 1])
 at [Source: (String)"{"serviceContext":{"service":"leonardo"}"; line: 1, column: 41]
        at com.fasterxml.jackson.core.io.JsonEOFException: Unexpected end-of-input: expected close marker for Object (start marker at [Source: (String)"{"serviceContext":{"service":"leonardo"}"; line: 1, column: 1])
 at [Source: (String)"{"serviceContext":{"service":"leonardo"}"; line: 1, column: 41]
        at      at com.fasterxml.jackson.core.base.ParserMinimalBase._reportInvalidEOF(ParserMinimalBase.java:682)
        at      at com.fasterxml.jackson.core.base.ParserBase._handleEOF(ParserBase.java:494)
        at      at com.fasterxml.jackson.core.base.ParserBase._eofAsNextChar(ParserBase.java
```

If I change `customFields` to ` {"serviceContext": "foo"}`, things work as expected.

I tried upgrading `logback-classic` to `1.3.0-alpha10`, but no log was generated

**Comment by brenuart** [comment]
Hi @Qi77Qi ,

Sorry to hear you have a problem.
As mentioned by @philsttr, this unfortunate behaviour is caused by a Logback bug ([LOGBACK-1101](http://jira.qos.ch/browse/LOGBACK-1101) which appears to be solved in Logback 1.2.0-alpha3.

I did a test with your configuration after upgrading to Logback 1.3.0-alpha10 and could not reproduce what you experience. The pattern string is now properly passed to the LogstashEncoder and not corrupted by Logstash's property substitution mechanism anymore. No exception is thrown anymore and the "serviceContext" custom field is properly added as last field of the generated JSON.

Could you please double check your setup and make sure your application is actually using the latest Logback? Do not hesitate to attach a fully working Maven project to demonstrate the issue if needed...
